tsl/src/continuous_aggs/create.c

```diff
@@ -1180,25 +1185,141 @@ cagg_validate_query(const Query *query, const bool finalized, const char *cagg_s
-	if (list_length(fromList) != 1 || !IsA(linitial(fromList), RangeTblRef))
+	if (list_length(fromList) > CONTINUOUS_AGG_MAX_JOIN_RELATIONS)
 	{
 		ereport(ERROR,
 				(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
-				 errmsg("only one hypertable allowed in continuous aggregate view")));
+				 errmsg("only two tables with one hypertable and one normal table"
+						"are  allowed in continuous aggregate view")));
 	}
-	/* check if we have a hypertable in the FROM clause */
-	rtref = linitial_node(RangeTblRef, query->jointree->fromlist);
-	rte = list_nth(query->rtable, rtref->rtindex - 1);
+	/* Extra checks for joins in Caggs */
+	if (list_length(fromList) == CONTINUOUS_AGG_MAX_JOIN_RELATIONS ||
+		!IsA(linitial(query->jointree->fromlist), RangeTblRef))
+	{
+		/* Using old format caggs is not supported */
+		if (!finalized)
+			ereport(ERROR,
+					(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+					 errmsg("old format of continuous aggregate is not supported with joins"),
+					 errhint("set timescaledb.finalized to TRUE")));
+
+		if (list_length(fromList) == CONTINUOUS_AGG_MAX_JOIN_RELATIONS)
+		{
+			if (!IsA(linitial(fromList), RangeTblRef) || !IsA(lsecond(fromList), RangeTblRef))
+				ereport(ERROR,
+						(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+						 errmsg("invalid continuous aggregate view"),
+						 errdetail(
+							 "from clause can only have one hypertable and one normal table")));
+
+			rtref = linitial_node(RangeTblRef, query->jointree->fromlist);
+			rte = list_nth(query->rtable, rtref->rtindex - 1);
+			rtref_other = lsecond_node(RangeTblRef, query->jointree->fromlist);
+			rte_other = list_nth(query->rtable, rtref_other->rtindex - 1);
+			jointype = rte->jointype || rte_other->jointype;
+
+			if (query->jointree->quals != NULL && IsA(query->jointree->quals, OpExpr))
+				op = (OpExpr *) query->jointree->quals;
+		}
+		else
+		{
+			ListCell *l;
+			foreach (l, query->jointree->fromlist)
+			{
+				Node *jtnode = (Node *) lfirst(l);
+				JoinExpr *join = NULL;
+				if (IsA(jtnode, JoinExpr))
+				{
+					join = castNode(JoinExpr, jtnode);
+#if PG13_LT
+					if (join->usingClause != NULL)
+						ereport(ERROR,
+								(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+								 errmsg("invalid continuous aggregate view"),
+								 errdetail(
+									 "joins with using clause in continuous aggregate definition"
+									 " work for postgres versions 13 and above")));
+#endif
+					jointype = join->jointype;
+					op = (OpExpr *) join->quals;
+					rte = list_nth(query->rtable, ((RangeTblRef *) join->larg)->rtindex - 1);
+					rte_other = list_nth(query->rtable, ((RangeTblRef *) join->rarg)->rtindex - 1);
+				}
+			}
+		}
+
+		/*
+		 * Cagg with joins does not support nested caggs in from clause.
+		 */
+		if (rte->relkind == RELKIND_VIEW || rte_other->relkind == RELKIND_VIEW)
+			ereport(ERROR,
+					(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+					 errmsg("joins for nested continuous aggregates are not supported")));
```

Comment:
```suggestion
					 errmsg("joins for hierarchical continuous aggregates are not supported")));
```
Forgot to add a comment here before but we should use "Hierarchical" instead of "Nested" when messaging to the users because this was the official name of the feature. See our docs: https://docs.timescale.com/timescaledb/latest/how-to-guides/continuous-aggregates/hierarchical-continuous-aggregates/